### PR TITLE
feature (refs T34551): refactor procedure customer relation

### PR DESCRIPTION
**Pull request by MoritzMandler**
Ticket:
https://yaits.demos-deutschland.de/T34551

Description:
alters the property customer for the procedure table
This property was used to indicate a blueprint that is just for a specific customer
Now this property shall indicate that the procedure is owned by an organisation within this customer
This is necessary to determine all customer related activity.
The logic to filter a customer-blueprint had to be adjusted in order to restore the previously expected behavior.

### How to review/test
check for instance diplanbau -> login as a 'Mandanten-Administration' -> goTo 'Blaupause' 
and check the results and behavior stays the same  as compared to main... (on diplanbau you have the option to set a default-customer-blueprint)
check ewm -> do the same (ewm does not have the option to edit a default-customer-blueprint)

you can check the migrations up/down via:
```bin/ewm doc:mig:exe Application\\Migrations\\Version20240105113406 --up```
```bin/ewm doc:mig:exe Application\\Migrations\\Version20240108110342 --up```
the second one gives you an echo output for organisations that are Kommune or Anhörungsbehörde in multiple customers
or both types within the same customer. These are expceted to be only our own test-organisations!!!  example output:
 ```
Hannes1 für AG - is "Kommune or Anhörungsbehörde" in customers: sh, sh
Amt Nordwest - is "Kommune or Anhörungsbehörde" in customers: sh, by
```
The first example row indicates the organisation is Komnuna as well as Anhörungsbehörde in sh - which is completely fine as the customer to set is the same.
The second row would not be fine if it wouldn't be our own testorganisation - the customer to set is unclear and so the first one will be chosen. 

### Linked PRs (optional)
demosplan-addon: https://github.com/demos-europe/demosplan-addon/pull/85

checked for usages of changed methods using the github searchBar with ```"isCustomerMasterBlueprint" AND (repo:demos-europe/demosplan-addon-xbeteiligung-async OR repo:demos-europe/demosplan-addon-segment-oracle OR repo:demos-europe/demosplan-addon-demospipes OR repo:demos-europe/demosplan-addon-maillane OR repo:demos-europe/demosplan-addon-datasheets-sh OR repo:demos-europe/demosplan-addon-bimschg-antrag)``` for ```isCustomerMasterBlueprint and setCustomer``` - no usages found.

- [ ] Tests updated/created
- [ ] Update documentation
- [x] Link all relevant tickets
- [ ] Move the tickets on the board accordingly

### CODE REVIEW

File: demosplan/DemosPlanCoreBundle/DoctrineMigrations/2024/01/Version20240108110342.php [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+class Version20240108110342 extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return 'refs T34551: alters the property customer for the procedure table
+            This property was used to indicate a blueprint that is just for a specific customer
+            Now this property shall indicate that the procedure is owned by an organisation within this customer
+            This is necessary to determine all customer related activity.
+            The logic to filter a customer-blueprint has to be adjusted to check the additional property _p_master
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/DoctrineMigrations/2024/01/Version20240108110342.php:28]: Note to self for the review: Check this one that it does not break the "usual" orga procedure template logic

File: demosplan/DemosPlanCoreBundle/DoctrineMigrations/2024/01/Version20240108110342.php [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        $customerKommunes = $this->selectCustomerKommunes('accepted');
+        // List Organisations that are accepted as "Kommune" in more than one customer
+        // these are expected to be our test-Organisations!
+        echo $this->transformKommunesOfMultipleCustomerIntoReadableOutput(
+            $this->getKommunesOfMultipleCustomers($customerKommunes)
+        );
+        // proceed setting the customer property for procedures but do not overwrite existing relations
+        $this->setCustomerRelations($customerKommunes);
+        // if there are procedures left without customer relation - try to set the leftovers with now pending kommunes/Ahbs
+        $pendingCustomerKommunes = $this->selectCustomerKommunes('pending');
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/DoctrineMigrations/2024/01/Version20240108110342.php:59]: Pending Orgas could not create procedures. Maybe this can be removed?

File: demosplan/DemosPlanCoreBundle/DoctrineMigrations/2024/01/Version20240108110342.php [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        $this->setCustomerRelations($pendingCustomerKommunes);
+        // if there are procedures left without customer relation - try to set the leftovers with now rejected kommunes/Ahbs
+        $rejectedCustomerKommunes = $this->selectCustomerKommunes('rejected');
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/DoctrineMigrations/2024/01/Version20240108110342.php:62]: Same here

File: demosplan/DemosPlanCoreBundle/Entity/Procedure/Procedure.php [hunk 4]
```diff
@@ -2006,11 +2006,6 @@ public function getSubdomain(): string
         return $customer->getSubdomain();
     }
 
-    public function isCustomerMasterBlueprint(): bool
-    {
-        return $this->master && null !== $this->customer;
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Entity/Procedure/Procedure.php:2011]: see above. I do not see why this should be removed as we still need the information

File: demosplan/DemosPlanCoreBundle/Entity/Procedure/Procedure.php [hunk 5]
```diff
@@ -2029,20 +2024,9 @@ public function getCustomerId()
 
     /**
      * @param CustomerInterface|null $customer
-     * @param bool                   $handleBothSites
      */
-    public function setCustomer($customer, $handleBothSites = true): Procedure
+    public function setCustomer($customer): Procedure
     {
-        if ($handleBothSites) {
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Entity/Procedure/Procedure.php:2036]: Note to self for review: How is the defaultProcedureBlueprint now set

File: demosplan/DemosPlanCoreBundle/Entity/User/Customer.php [hunk 6]
```diff
@@ -113,7 +113,7 @@ class Customer extends CoreEntity implements UuidEntityInterface, CustomerInterf
      *
      * @var ProcedureInterface
      *
-     * @ORM\OneToOne(targetEntity="demosplan\DemosPlanCoreBundle\Entity\Procedure\Procedure", mappedBy="customer")
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Entity/User/Customer.php:116]: Why did you remove `mappedBy="customer"`?

File: demosplan/DemosPlanCoreBundle/Entity/Procedure/Procedure.php [hunk 7]
```diff
@@ -515,7 +515,7 @@ class Procedure extends SluggedEntity implements ProcedureInterface
      *
      * @var Customer
      *
-     * @ORM\OneToOne(targetEntity="demosplan\DemosPlanCoreBundle\Entity\User\Customer", inversedBy="defaultProcedureBlueprint", cascade={"remove", "persist"})
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Entity/Procedure/Procedure.php:518]: Why did you remove `inversedBy="defaultProcedureBlueprint"`?

File: demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureService.php [hunk 8]
```diff
@@ -828,14 +842,36 @@ public function addProcedureEntity(array $data, string $currentUserId): Procedur
                 $data['settings']['emailTitle'] =
                     $this->translator->trans('participation.invitation').': '.($data['name'] ?? '');
             }
-
+            // T34551 all procedures shall get a customer relation
+            // - defalt-customer-blueprint relations are set within the customer only
+            // if a customer is given inside the procedure related $data array then
+            // that signals the procedure should be used as the default-customer-blueprint.
+            $setProcedureAsDefaultCustomerBlueprint = false;
+            if (array_key_exists('custoemr', $data) && $data['customer'] instanceof Customer) {
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureService.php:850]: typo in custoemr. Did you test this case?

File: demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureService.php [hunk 9]
```diff
@@ -2112,7 +2148,9 @@ protected function setAuthorizedUsersToProcedure(Procedure $newProcedure, $bluep
             // T15644: T23583:
             // overwrite authorized users in case of used blueprint is a master-blueprint,
             // to avoid authorizing creators of masterblueprint to this new procedure:
-            if ($blueprint->isMasterTemplate() || $blueprint->isCustomerMasterBlueprint()) {
+            if ($blueprint->isMasterTemplate()
+                || $blueprint->getId() === $blueprint->getCustomer()?->getDefaultProcedureBlueprint()?->getId()
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureService.php:2152]: I would keep this check in `isCustomerMasterBlueprint()`

File: demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureToLegacyConverter.php [hunk 10]
```diff
@@ -102,16 +103,23 @@ protected function convertOrgasToLegacy(Procedure $procedure): array
     {
         $nonPlanningOfficeOrganisationIds = $this->procedureRepository->getInvitedOrgaIds($procedure->getId());
         $planningOfficeIds = $this->procedureRepository->getPlanningOfficeIds($procedure->getId());
-        $isCustomerMasterBlueprint = $procedure->isCustomerMasterBlueprint();
+        $isCustomerMasterBlueprint =
+            $procedure->getId() === $procedure->getCustomer()?->getDefaultProcedureBlueprint()?->getId();
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureToLegacyConverter.php:107]: See above, keep method

File: demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureToLegacyConverter.php [hunk 11]
```diff
@@ -102,16 +103,23 @@ protected function convertOrgasToLegacy(Procedure $procedure): array
...
                 'name'      => $orga->getName(),
                 'nameLegal' => $orga->getName(),
             ])->all();
         $dataInputOrgaIds = $procedure->getDataInputOrgaIds();
         $authorizedUserIds = $procedure->getAuthorizedUserIds();
 
+        // T34551 changed the relation of procedure to customer.
+        // previously the procedure->customer relation was null Except for the default-customer-blueprint.
+        // Now only the customer holds the relation to its default-blueprint.
+        $customerToLegacy = $procedure->getId() === $procedure->getCustomer()?->getDefaultProcedureBlueprint()?->getId() ?
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureToLegacyConverter.php:120]: use `isCustomerMasterBlueprint`

File: demosplan/DemosPlanCoreBundle/Logic/Procedure/ServiceStorage.php [hunk 12]
```diff
@@ -360,19 +350,25 @@ public function administrationEditHandler($data, $checkMandatoryErrors = true)
...
-                );
+        if (!$isMasterTemplate) {
+            $procedure['customer'] = $currentCustomer;
+            if (array_key_exists('r_customerMasterBlueprint', $data)) {
+                $currentCustomer->setDefaultProcedureBlueprint($currentProcedure);
+                $this->customerService->updateCustomer($currentCustomer);
             } else {
-                $procedure['customer'] = $currentCustomer;
+                // T15644 & T34551 if the key 'r_customerMasterBlueprint' is not set within the $data array,
+                // - the assumpten is tha the procedure shall not be the default-customer-blueprint
```
Comment (mussbach) [inline demosplan/DemosPlanCoreBundle/Logic/Procedure/ServiceStorage.php:364]:
```suggestion
                // - the assumption is that the procedure shall not be the default-customer-blueprint
```

**Review by mussbach — CHANGES_REQUESTED** [review]

**Review by bakridemos — CHANGES_REQUESTED** [review]

File: demosplan/DemosPlanCoreBundle/Controller/Procedure/DemosPlanProcedureController.php [hunk 13]
```diff
@@ -1417,7 +1417,8 @@ public function administrationEditAction(
                 });
             }
 
-            $templateVars['isCustomerMasterBlueprint'] = $procedureObject->isCustomerMasterBlueprint();
+            $templateVars['isCustomerMasterBlueprint'] =
+                $procedureObject->getId() === $procedureObject->getCustomer()?->getDefaultProcedureBlueprint()?->getId();
```
[the remark this answers is not in the document]
Comment (MoritzMandler) [reply demosplan/DemosPlanCoreBundle/Controller/Procedure/DemosPlanProcedureController.php]: restored in 6f731c5800aeab34e5de2da82e2fb5c1a96e133c

File: demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureService.php (hunk 9 above)
Comment (MoritzMandler) [reply demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureService.php]: restored in 6f731c5800aeab34e5de2da82e2fb5c1a96e133c

File: demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureToLegacyConverter.php (hunk 10 above)
Comment (MoritzMandler) [reply demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureToLegacyConverter.php]: restored in 6f731c5800aeab34e5de2da82e2fb5c1a96e133c

File: demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureToLegacyConverter.php (hunk 11 above)
Comment (MoritzMandler) [reply demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureToLegacyConverter.php]: restored in 6f731c5800aeab34e5de2da82e2fb5c1a96e133c

File: demosplan/DemosPlanCoreBundle/Entity/Procedure/Procedure.php (hunk 4 above)
Comment (MoritzMandler) [reply demosplan/DemosPlanCoreBundle/Entity/Procedure/Procedure.php:2011]: restored in 6f731c5800aeab34e5de2da82e2fb5c1a96e133c

File: demosplan/DemosPlanCoreBundle/DoctrineMigrations/2024/01/Version20240105113406.php [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Platforms\MySQLPlatform;
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+class Version20240105113406 extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return 'refs TT34551: Remove unique-constraint for procedures - customer
```
[the remark this answers is not in the document]
Comment (MoritzMandler) [reply demosplan/DemosPlanCoreBundle/DoctrineMigrations/2024/01/Version20240105113406.php]: done in a4ca32805b517f67d9bbc2190a511aed3f21da31

File: demosplan/DemosPlanCoreBundle/DoctrineMigrations/2024/01/Version20240108110342.php [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Platforms\MySQLPlatform;
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+class Version20240108110342 extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return 'refs T34551: alters the property customer for the procedure table
```
[the remark this answers is not in the document]
Comment (MoritzMandler) [reply demosplan/DemosPlanCoreBundle/DoctrineMigrations/2024/01/Version20240108110342.php]: done in a4ca32805b517f67d9bbc2190a511aed3f21da31

File: demosplan/DemosPlanCoreBundle/Entity/User/Customer.php (hunk 6 above)
Comment (MoritzMandler) [reply demosplan/DemosPlanCoreBundle/Entity/User/Customer.php:116]: This was a bi-directional one-to-one relation.
Now there are two uni-directional one-to-one relations.
procedure -> holds reference to its customer
customer -> holds reference to its default-blueprint-procedure

File: demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureService.php (hunk 8 above)
Comment (MoritzMandler) [reply demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureService.php]: This method gets only called when creating a new bluePrint.
As a 'Plattform-Administration' there is a checkbox to mark the blueprint as default.
Interestingly there is never the key 'customer' present and there is simply no option to detect the default-customer-blueprint checkbox was active when creating a NEW procedure.
To set the blueprint as default - an UPDATE has to be triggered.

Fixed the typo - but it has no effect. 2abd20044a336d28b24666c75d053c014998bd18

[the remark this answers is not in the document]
Comment (MoritzMandler) [reply demosplan/DemosPlanCoreBundle/Logic/Procedure/ProcedureService.php]: This method gets only called when creating a new bluePrint.
As a 'Plattform-Administration' there is a checkbox to mark the blueprint as default.
Interestingly there is never the key 'customer' present and there is simply no option to detect the default-customer-blueprint checkbox was active when creating a NEW procedure.
To set the blueprint as default - an UPDATE has to be triggered.

Fixed the typo - but it has no effect. https://github.com/demos-europe/demosplan-core/commit/2abd20044a336d28b24666c75d053c014998bd18

**Review by bakridemos — DISMISSED** [review]
lgtm. code review and tested too
